File: packages/shared/src/types/cron.ts
import { z } from 'zod';

// ============================================================================
// Cron Schedule Types
// ============================================================================

/**
 * One-shot schedule: runs once at a specific timestamp
 */
export const CronScheduleAtSchema = z.object({
  type: z.literal('at'),
  timestamp: z.number().describe('UTC timestamp in milliseconds'),
});

/**
 * Recurring interval schedule: runs every N milliseconds
 */
export const CronScheduleEverySchema = z.object({
  type: z.literal('every'),
  intervalMs: z.number().min(60000).describe('Interval in milliseconds (min 1 minute)'),
  startAt: z.number().optional().describe('Optional start timestamp for alignment'),
});

/**
 * Cron expression schedule: standard cron format
 */
export const CronScheduleCronSchema = z.object({
  type: z.literal('cron'),
  expression: z.string().describe('Cron expression (e.g., "0 9 * * MON-FRI")'),
  timezone: z.string().optional().describe('IANA timezone (e.g., "America/Los_Angeles")'),
});

/**
 * Discriminated union of all schedule types
 */
export const CronScheduleSchema = z.discriminatedUnion('type', [
  CronScheduleAtSchema,
  CronScheduleEverySchema,
  CronScheduleCronSchema,
]);

export type CronSchedule = z.infer<typeof CronScheduleSchema>;
export type CronScheduleAt = z.infer<typeof CronScheduleAtSchema>;
export type CronScheduleEvery = z.infer<typeof CronScheduleEverySchema>;
export type CronScheduleCron = z.infer<typeof CronScheduleCronSchema>;

// ============================================================================
// Cron Job Types
// ============================================================================

/**
 * Session target determines where the job runs
 */
export const CronSessionTargetSchema = z.enum(['main', 'isolated']);
export type CronSessionTarget = z.infer<typeof CronSessionTargetSchema>;

/**
 * Job status
 */
export const CronJobStatusSchema = z.enum(['active', 'paused', 'completed', 'failed']);
export type CronJobStatus = z.infer<typeof CronJobStatusSchema>;

/**
 * Wake mode for main session jobs
 */
export const CronWakeModeSchema = z.enum(['next-heartbeat', 'now']);
export type CronWakeMode = z.infer<typeof CronWakeModeSchema>;

/**
 * Last run status
 */
export const CronLastStatusSchema = z.enum(['ok', 'error', 'skipped']);
export type CronLastStatus = z.infer<typeof CronLastStatusSchema>;

/**
 * Core cron job definition
 */
export const CronJobSchema = z.object({
  id: z.string(),
  name: z.string().min(1).max(100),
  description: z.string().max(500).optional(),
  prompt: z.string().min(1).max(10000),
  schedule: CronScheduleSchema,
  sessionTarget: CronSessionTargetSchema,
  wakeMode: CronWakeModeSchema,
  workingDirectory: z.string(),
  model: z.string().optional(),
  status: CronJobStatusSchema,
  deleteAfterRun: z.boolean().optional(),
  maxRuns: z.number().int().positive().optional(),
  tags: z.array(z.string()).optional(),

  // Timestamps
  createdAt: z.number(),
  updatedAt: z.number(),

  // State
  runCount: z.number().int().nonnegative(),
  lastRunAt: z.number().optional(),
  nextRunAt: z.number().optional(),
  lastStatus: CronLastStatusSchema.optional(),
  lastError: z.string().optional(),
  lastDurationMs: z.number().optional(),
});

export type CronJob = z.infer<typeof CronJobSchema>;

// ============================================================================
// Cron Run Types
// ============================================================================

/**
 * Run result status
 */
export const CronRunResultSchema = z.enum(['success', 'error', 'timeout', 'cancelled']);
export type CronRunResult = z.infer<typeof CronRunResultSchema>;

/**
 * Run history entry (stored as JSONL)
 */
export const CronRunSchema = z.object({
  id: z.string(),
  jobId: z.string(),
  sessionId: z.string(),
  startedAt: z.number(),
  completedAt: z.number().optional(),
  result: CronRunResultSchema,
  error: z.string().optional(),
  summary: z.string().optional(),
  promptTokens: z.number().optional(),
  completionTokens: z.number().optional(),
});

export type CronRun = z.infer<typeof CronRunSchema>;

// ============================================================================
// Cron Service Types
// ============================================================================

/**
 * Tool actions for agent tool
 */
export const CronToolActionSchema = z.enum([
  'status',
  'list',
  'add',
  'update',
  'remove',
  'pause',
  'resume',
  'run',
  'runs',
  'wake',
]);

export type CronToolAction = z.infer<typeof CronToolActionSchema>;

/**
 * Cron service status
 */
export const CronServiceStatusSchema = z.object({
  isRunning: z.boolean(),
  jobCount: z.number(),
  activeJobCount: z.number(),
  nextRunAt: z.number().nullable(),
  nextJobId: z.string().nullable(),
});

export type CronServiceStatus = z.infer<typeof CronServiceStatusSchema>;

// ============================================================================
// Input Types (for API/IPC)
// ============================================================================

/**
 * Create job input (without auto-generated fields)
 */
export const CreateCronJobInputSchema = z.object({
  name: z.string().min(1).max(100),
  description: z.string().max(500).optional(),
  prompt: z.string().min(1).max(10000),
  schedule: CronScheduleSchema,
  sessionTarget: CronSessionTargetSchema,
  wakeMode: CronWakeModeSchema,
  workingDirectory: z.string(),
  model: z.string().optional(),
  deleteAfterRun: z.boolean().optional(),
  maxRuns: z.number().int().positive().optional(),
  tags: z.array(z.string()).optional(),
});

export type CreateCronJobInput = z.infer<typeof CreateCronJobInputSchema>;

/**
 * Update job input (all fields optional)
 */
export const UpdateCronJobInputSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().max(500).optional(),
  prompt: z.string().min(1).max(10000).optional(),
  schedule: CronScheduleSchema.optional(),
  sessionTarget: CronSessionTargetSchema.optional(),
  wakeMode: CronWakeModeSchema.optional(),
  workingDirectory: z.string().optional(),
  model: z.string().optional(),
  status: z.enum(['active', 'paused']).optional(),
  deleteAfterRun: z.boolean().optional(),
  maxRuns: z.number().int().positive().optional(),
  tags: z.array(z.string()).optional(),
});

export type UpdateCronJobInput = z.infer<typeof UpdateCronJobInputSchema>;

/**
 * Run query options
 */
export const RunQueryOptionsSchema = z.object({
  limit: z.number().int().positive().optional(),
  offset: z.number().int().nonnegative().optional(),
  result: CronRunResultSchema.optional(),
});

export type RunQueryOptions = z.infer<typeof RunQueryOptionsSchema>;

File: apps/desktop/src-tauri/src/commands/cron.rs
use crate::commands::agent::{ensure_sidecar_started_public, AgentState};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, State};

// ============================================================================
// Cron Schedule Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum CronSchedule {
    #[serde(rename = "at")]
    At { timestamp: i64 },
    #[serde(rename = "every")]
    Every {
        interval_ms: i64,
        #[serde(skip_serializing_if = "Option::is_none")]
        start_at: Option<i64>,
    },
    #[serde(rename = "cron")]
    Cron {
        expression: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        timezone: Option<String>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CronJob {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub prompt: String,
    pub schedule: CronSchedule,
    pub session_target: String, // "main" | "isolated"
    pub wake_mode: String,      // "next-heartbeat" | "now"
    pub working_directory: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    pub status: String, // "active" | "paused" | "completed" | "failed"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delete_after_run: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_runs: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    pub created_at: i64,
    pub updated_at: i64,
    pub run_count: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_run_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_run_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_duration_ms: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CronRun {
    pub id: String,
    pub job_id: String,
    pub session_id: String,
    pub started_at: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<i64>,
    pub result: String, // "success" | "error" | "timeout" | "cancelled"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completion_tokens: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCronJobInput {
    pub name: String,
    pub prompt: String,
    pub schedule: CronSchedule,
    pub working_directory: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_target: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wake_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delete_after_run: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_runs: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateCronJobInput {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub schedule: Option<CronSchedule>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_target: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wake_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delete_after_run: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_runs: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

// ============================================================================
// Tauri Commands
// ============================================================================

/// List all cron jobs
#[tauri::command]
pub async fn cron_list_jobs(
    app: AppHandle,
    state: State<'_, AgentState>,
) -> Result<Vec<CronJob>, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let result = manager
        .send_command("cron_list_jobs", serde_json::json!({}))
        .await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse cron jobs: {}", e))
}

/// Get a specific cron job by ID
#[tauri::command]
pub async fn cron_get_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
) -> Result<CronJob, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::json!({ "jobId": job_id });
    let result = manager.send_command("cron_get_job", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse cron job: {}", e))
}

/// Create a new cron job
#[tauri::command]
pub async fn cron_create_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    input: CreateCronJobInput,
) -> Result<CronJob, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::to_value(&input)
        .map_err(|e| format!("Failed to serialize input: {}", e))?;
    let result = manager.send_command("cron_create_job", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse created job: {}", e))
}

/// Update an existing cron job
#[tauri::command]
pub async fn cron_update_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
    input: UpdateCronJobInput,
) -> Result<CronJob, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let mut params = serde_json::to_value(&input)
        .map_err(|e| format!("Failed to serialize input: {}", e))?;
    params["jobId"] = serde_json::json!(job_id);
    let result = manager.send_command("cron_update_job", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse updated job: {}", e))
}

/// Delete a cron job
#[tauri::command]
pub async fn cron_delete_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
) -> Result<(), String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::json!({ "jobId": job_id });
    manager.send_command("cron_delete_job", params).await?;

    Ok(())
}

/// Pause a cron job
#[tauri::command]
pub async fn cron_pause_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
) -> Result<CronJob, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::json!({ "jobId": job_id });
    let result = manager.send_command("cron_pause_job", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse job: {}", e))
}

/// Resume a paused cron job
#[tauri::command]
pub async fn cron_resume_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
) -> Result<CronJob, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::json!({ "jobId": job_id });
    let result = manager.send_command("cron_resume_job", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse job: {}", e))
}

/// Trigger immediate execution of a cron job
#[tauri::command]
pub async fn cron_trigger_job(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
) -> Result<CronRun, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::json!({ "jobId": job_id });
    let result = manager.send_command("cron_trigger_job", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse run result: {}", e))
}

/// Get run history for a cron job
#[tauri::command]
pub async fn cron_get_runs(
    app: AppHandle,
    state: State<'_, AgentState>,
    job_id: String,
    limit: Option<u32>,
    offset: Option<u32>,
) -> Result<Vec<CronRun>, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    let params = serde_json::json!({
        "jobId": job_id,
        "limit": limit,
        "offset": offset,
    });
    let result = manager.send_command("cron_get_runs", params).await?;

    serde_json::from_value(result).map_err(|e| format!("Failed to parse runs: {}", e))
}

/// Get cron service status
#[tauri::command]
pub async fn cron_get_status(
    app: AppHandle,
    state: State<'_, AgentState>,
) -> Result<serde_json::Value, String> {
    ensure_sidecar_started_public(&app, &state).await?;

    let manager = &state.manager;
    manager
        .send_command("cron_get_status", serde_json::json!({}))
        .await
}
